ops/pebble.py

```diff
@@ -1926,3 +1910,198 @@ def _websocket_url(self, task_id: str, websocket_id: str) -> str:
         base_url = self.base_url.replace('http://', 'ws://')
         url = '{}/v1/tasks/{}/websocket/{}'.format(base_url, task_id, websocket_id)
         return url
+
+
+class MultipartLargeFileParser:
+    """A limited purpose multi-part parser backed by files for memory efficiency."""
+
+    def __init__(self, boundary: bytes):
+        self._buffer = b''
+        self._response = None
+        self._files = {}
+
+        # Store the initial FSM state.  See _run_parser_fsm() for a full explanation.
+        self._initial_state_fn = self._state_detect_boundary
+
+        # Prepare the MIME multipart boundary line matcher.
+        # Note: RFC 2046 notes optional "linear whitespace" (e.g. [ \t]) after the boundary pattern
+        # and the optional "--" suffix.  Pebble doesn't use this, so to simplify implementation I
+        # am ignoring it.
+        boundary_rgx_pattern = r"--{}(--)?\r\n".format(boundary)
+        self._boundary_rgx = re.compile(boundary_rgx_pattern.encode())
+
+        # When streaming file data to tempfiles, we need to make sure we don't feed part of the
+        # boundary line by mistake.  This is easy if we detect the boundary in full, but if we
+        # haven't received it yet, we need to be careful not to feed any partial boundary line
+        # to disk.  This minimum buffer length threshold will ensure that, and will work for both
+        # regular and terminating boundary cases.
+        #
+        # Calculated as: boundary header length + prefix/suffix -- (4 bytes) + \r\n (2 bytes)
+        self._safe_buffer_len = len(boundary) + 6
+
+    def __del__(self):
+        for file in self._files.values():
```

Comment:
I don't know if `__del__` acts quite as nicely as we would like it to.
Specifically, I think that `MultipartLargeFileParser` creates a reference counting loop with itself because it saves references to its methods, which means you have a refcount to a method which then has a refcount to the object.
```
>>> class Foo:
...   def __init__(self):
...     self._next = self._do_next
...   def _do_next(self):
...     print("next")
...   def __del__(self):
...     print("__del__")
...
>>> f = Foo()
>>> del f
>>> import gc
>>> gc.collect()
__del__
13
```
vs
```
>>> class Bar:
...   def __init__(self):
...     pass
...   def __del__(self):
...     print("Bar.__del__")
...
>>> b = Bar()
>>> del b
Bar.__del__
```

I don't know that there is a clean way to do this (at least __del__ isn't releasing the last references to MLFP, and MLFP doesn't have references to other objects, so I think it will trigger, but we *might* want a `MLFP.close()` which removes self._initial_state_fn.
(Though also all of this can be ignored if we move away from a tempfile we have to clean up.)

Reply:
I agree that this feels wrong. I would rather see an explicit call to clean up the temp file, if that's what we want to do, rather than relying on filesystem cleanup to happen as effectively a side effect of Python's garbage collector.

Reply:
Yep, it's typically to have an explicit `close()` method for things like this, or add `with` statement support by defining `__enter__` and `__exit__` (though the latter might be a bit heavy here).

Reply:
There are actually two issues here: closing the file, and removing the temporary file.  In order to facilitate the user getting an open file with their own mode/encoding params different from those used when the parser is writing the file, we need to close+re-open before we hand the object/file back to the caller of `pull`.  Because of this, the tempfile's auto-delete-on-close has been disabled.  If we want the charm-container to behave a bit more stateless by default, then it would be nice to auto-remove this temporary file upon close.  And the caller would be free to copy the data into a non-temp location if they want to persist it.  But if the "normal" use case is to have the file persist, then we should just kill this `__del__` code.  To accomplish the former - we basically would need to use something like the tempfile lib does internally with its `_TemporaryFileWrapper` class.  I'm new enough to this project that I'm not sure which behavior is more natural/preferrable.

Reply:
I don't think we want our copy of the content buffer to exist, as it will never be in the right location for the caller.

Reply:
Gustavo made a good suggestion (which I've implemented).  Before we return the (opened) file to the caller of `pull`, we can remove the file from disk.  The data will remain available until the file is closed (by the user or the gc).  This way we don't introduce any new API (e.g. a file on disk the user can do stuff with) - and limit them to reading from returned object directly.  We can continue to ruminate on temp file locations, copying, etc. details until we are confident if/how we want to change things.  I'm going to consider this resolved for now.